Standalone/mkftl.1
.TH MKFTL 1 2021-03-03 "mkftl manual"

.SH NAME
mkftl \- create Dhara FTL images

.SH SYNOPSIS
.B mkftl [\fIOPTION\fR] <\fIinputfile.img\fR> -o <\fIoutputfile.ftl\fR>

.SH DESCRIPTION

\fImkftl\fR is a program which converts normal images into FTL images used by
the Dhara FTL layer. It's useful for creating images which can be written to
NAND devices.

.SH OPTIONS
\fImkftl\fR accepts the following options:

.TP
.BR \-p " size in bytes"
Specifies the page size of the Dhara map. Default: 512 bytes.

.TP
.BR \-e " size in bytes"
Specifies the erase block size of the Dhara map. Default: 4096 bytes.

.TP
.BR \-g " ratio"
Specifies the garbage collection ratio of the Dhara map. Default: 2. The larger
the number, the less waste there will be in the map, but the more garbage
collection will be done on each write.

.TP
.BR \-s " size in kilobytes"
Specifies the physical size of the Dhara map. The bigger it is, the bigger a
logical image can be contained within it; however, if the logical image
completely fills the map, Dhara will have trouble operating on it and will
probably perform garbage collection thrashing.

The filesystem parameters must match those of the device which will be using
it, or else you'll get random corruption and crashes.

.SH SEE ALSO
https://github.com/dlbeer/dhara



RELEASE-NOTES

			Fuzix 0.3 Release Notes

Fuzix 0.3 is the third major release of the Fuzix OS. This release has mostly
been focussed upon kernel improvements and supporting more platforms,
particularly ZX Spectrum derived ones. The 68000 code base is now usable
but not entirely debugged and optimized. 8080 and 8085 support has been added.

Documentation

There is now a documentation directory and rules to build it
into something. It's still very incomplete and badly formatted

Warning

Fuzix 0.4 will break ABI compatibility big time. There are a pile of bad
and historic design decisions and behaviours that need to be dealt with.


User Space Changes:

New Commands: tty, vile, sok, fweeplet (Zork engine), fsh (sh with editing)
Main Improved commands:
	cp,mv,ln now taken from Heirloom so full Unix style commands
	dd can use stdin/stdout
	fweep now generally available
	sh now correctly handles out of memory
	ucp now behaves correctly if you ls a file.
	sort is now based on a different smaller codebase.
	fsck check on a clean file system is much faster.


Behavioural Changes:
- Some systems had fd first not hd first. This has been unified so on
  platforms where 0 was fd0, it's now 256 and hda is 0. If you use the fd
  or hd names at boot prompts then it's not visible

Systems Added

Amstrad PCW 8256 with CF adapter (minimal early port)
Bill Shen
	SBC64/MBC64
	Simple80 (in test)
	Tiny68K (for 68K development - very glitchy still)
Cromemco with 16FDC and 8" disks
Easy-Z80
Grant Searle Z80 CP/M design with small modifications
Linc80 with extra 16K RAM card
Pentagon 1024 with NemoIDE
RC2014 with 6502 Processor (emulation only board debug in process)
RC2014 with 8085 Processor
Sam Coupe (minimal port only at this point), Atomlite IDE
Small Computer Central
	SC108	Z80 128K RAM, CF
	SC111	Z180, 512K RAM, CF
	SC114	Z80 128K RAM, CF
SBC2G (Grant Searle style system with banked memory)
Scorpion with NemoIDE
Scrumpel (Z180)
SmallZ80
Timex TC2068/TS2068 with DivIDE and Fuzix on a cartridge
Tom's SBC (with small mods)
Video Genie with EG64B banker (and in theory TRS80 + Lubomir soft banker)
Z80-MBC2
Z80 Membership card
ZX Spectrum 128K with DivIDE or DivMMC
ZX Spectrum +3 with ZXMMC or similar

Existing Supported Systems

Amstrad NC100 with 1MB SRAM card
Amstrad NC200 with 1MB SRAM card and floppy disk drive
Dragon 32/64 with Spinx or MOOH cartridge
DX Designs P112 with 1MB RAM and G-IDE
EACA Video Genie (with suitable banked memory)
LNW Research LNW80 with suitable banked memory (TRS80 Model 1 compatible bits)
Memotech MTX512 with SDX disk (no Rememorizer or Rememotech support)
Microbee 256TC or Premium with 128K+ RAM and ideally hard disk
Multicomp09 with at least 128K RAM
RBC(*) Mark IV with optional PropIO V2
RBC(*) SBC v2 with optional PPIDE and/or PropIO V2
RC2014 SBC with 512K ROM/RAM, CF, SIO and RTC
RC2014 SBC with banked ROM, 64K RAM, CF, SIO and RTC
SOCZ80 (128MHz FPGA Z80 platform)
Tandy COCO2 with Cloud 9 IDE, or COCOSDC card and Fuzix partly in cartridge
Tandy COCO3 with suitable disk interface
TRS80 Model I/III with a hard disk and a supported banked RAM expansion
TRS80 Model 4/4D/4P 128K RAM
Zeta
Zeta V2

(*) Formerly N8VEM now RBC (RetroBrewComputers)


Incomplete Ports

Apple IIc	-	initial investigations
Atari ST	-	in progress
EZRetro		-	minimal development port for ez80
Gemini		-	early design
Lucas Nascom	-	early design
MSX1		-	works in very limited configuration only
Pentagon	-	needs the 0.4 ABI changes to be possible
ZX Uno		-	for now run the 128K ZX Spectrum version as this also
			knows about some Uno features
Z80BIOS		-	experiment to see if things like S100 can be supported
			by having a Z80 BIOS akin to the CP/M BIOS etc.
Z280RC/ZZ80RC	-	early sketches


New Virtual Platforms (Emulation Platforms For Fun or Development)

v8080		-	8080 development environment
v85		-	8085 development environment

Existing Virtual Platforms

V65C816		-	65C816 emulation work
V68		-	68000 development work
Z80Pack		-	Very flexible Z80 banked system emulator

Obsolete

MSX2		-	needs a major rework and clean up

Major Changes To Existing Platforms

Dragon-MOOH	-	now it's own port with full MOOH paging support.
MTX		-	support for CFX-I CFX-II, some rememorize support,
			base MTXPlus support, extra keys on PC adapter. Lots
			of bug fixes.

The Library/Application support has been built for
6502			no FP
65C816			no FP, using 6502 modes
6809			FP needs debug
68000			FP needs debug
68HC11			test only
8086			test only
MSP430X
NS32k			test only
PDP11			test only
Rabbit2000		test only
Z80
Z180
eZ80			needs a very recent SDCC


Core Changes

- New memory banking model for 16K banks with a fixed common
- Timer handling optimized massively.
- Correct handling of platforms with an optional RTC when it is not present
  (previously clock would drift or not run)
- Correct problems if the console had carrier detect and could be hung up
- Fix races in carrier and hangup logic
- Clear error bits correctly on Z80 SIO
- Printer ioctls added
- 6502 platform support now complete including signal handling
- eZ80 support
- Z80 memory model support for machines with no usable common RAM
- Z80 switch to a software interrupt disable tracker rather than doing the
expensive NMOS Z80 workaround
- Z80 memory model support for 32K/32K split
- 8080/85 platform support complete
- 68000 flat memory model
- /dev/sys ioctl framework for platform / cpu specific ioctls
- Unified lots of drivewire code
- Support for "swapon"
- Single process in memory systems not run parent first after fork. That
  usually means that we hit waitpid() and avoids an extra set of swaps.
- Base support for Z180 CPU dropped into a Z80 environment
- Clock set from root fs superblock timestamp on start up. This avoids clocks
  going backwards and usually means the date is right when you boot and get
  asked.
- Use uint_fast8_t for platforms where uint8_t is slow, or the compiler sucks
  at handling them.
- Move repeated swap helper into partition code.

Standalone/mkftl.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>
#include <stdarg.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>
#include <errno.h>
#include "../Kernel/lib/dhara/map.h"
#include "../Kernel/lib/dhara/nand.h"

static uint32_t pagesize = 512;
static uint32_t erasesize = 4096;
static uint32_t flashsize = 1024*1024; /* bytes */
static uint8_t gcratio = 2;

static uint8_t* flashdata;

int dhara_nand_erase(const struct dhara_nand *n, dhara_block_t b,
                     dhara_error_t *err)
{
    memset(flashdata + (b*erasesize), 0xff, erasesize);
    *err = DHARA_E_NONE;
    return 0;
}

int dhara_nand_prog(const struct dhara_nand *n, dhara_page_t p,
                    const uint8_t *data,
                    dhara_error_t *err)
{
    memcpy(flashdata + (p*pagesize), data, pagesize);
    *err = DHARA_E_NONE;
    return 0;
}

int dhara_nand_read(const struct dhara_nand *n, dhara_page_t p,
                    size_t offset, size_t length,
                    uint8_t *data,
                    dhara_error_t *err)
{
    if (p >= (flashsize/pagesize))
    {
        *err = DHARA_E_BAD_BLOCK;
        return -1;
    }

    memcpy(data, flashdata + (p*pagesize) + offset, length);
    *err = DHARA_E_NONE;
    return 0;
}

int dhara_nand_is_bad(const struct dhara_nand* n, dhara_block_t b)
{
    return 0;
}

void dhara_nand_mark_bad(const struct dhara_nand *n, dhara_block_t b) {}

int dhara_nand_is_free(const struct dhara_nand *n, dhara_page_t p)
{
    const uint8_t* ptr = flashdata + (p*512);
    for (int i=0; i<512; i++)
        if (ptr[i] != 0xff)
            return 0;

    return 1;
}

int dhara_nand_copy(const struct dhara_nand *n,
                    dhara_page_t src, dhara_page_t dst,
                    dhara_error_t *err)
{
    const uint8_t* psrc = flashdata + (src*512);
    uint8_t* pdst = flashdata + (dst*512);
    memcpy(pdst, psrc, 512);
    *err = DHARA_E_NONE;
    return 0;
}

static void panic(const char* s, ...)
{
    va_list ap;
    va_start(ap, s);
    fprintf(stderr, "Fatal: ");
    vfprintf(stderr, s, ap);
    fprintf(stderr, "\n");
    va_end(ap);
    exit(1);
}

static bool is_power_of_2(uint32_t i)
{
    return i && !((i-1) & i);
}

static int ilog2(uint32_t i)
{
    return ffsl(i) - 1;
}

static void syntax_error(void)
{
    fprintf(stderr, "Syntax: mkftl [options] <inputfile> -o <outputfile>\n");
    fprintf(stderr, "Options:\n");
    fprintf(stderr, "  -p <number>   set page size, in bytes (default: %d bytes)\n", pagesize);
    fprintf(stderr, "  -e <number>   set erase block size, in bytes (default: %d bytes)\n", erasesize);
    fprintf(stderr, "  -s <number>   set image size, in kilobytes (default: %d kB)\n", flashsize/1024);
    fprintf(stderr, "  -g <number>   set garbage collection ratio (default: %d)\n", gcratio);
    exit(1);
}

int main(int argc, char* const* argv)
{
    const char* outputfilename = NULL;
    const char* inputfilename = NULL;

    for (;;)
    {
        int opt = getopt(argc, argv, "o:p:e:s:g:");
        if (opt == -1)
            break;
        switch (opt)
        {
            case 'o':
                outputfilename = optarg;
                break;

            case 'p':
                pagesize = strtoul(optarg, NULL, 0);
                break;

            case 'e':
                erasesize = strtoul(optarg, NULL, 0);
                break;

            case 's':
                flashsize = strtoul(optarg, NULL, 0) * 1024;
                break;

            case 'g':
                gcratio = strtoul(optarg, NULL, 0);
                break;

            default:
                syntax_error();
        }
    }

    if (argc != (optind+1))
        syntax_error();
    inputfilename = argv[optind];

    if (!outputfilename)
        panic("output filename must be supplied");

    if (!is_power_of_2(pagesize))
        panic("page size %d is not a power of two", pagesize);

    if (!is_power_of_2(erasesize))
        panic("erase size %d is not a power of two", erasesize);

    if (erasesize % pagesize)
        panic("erase size is not a multiple of page size");

    if (flashsize % erasesize)
        panic("flash size is not a multiple of erase size");

    struct dhara_nand nand;
    nand.log2_page_size = ilog2(pagesize);
    nand.log2_ppb = ilog2(erasesize) - nand.log2_page_size;
    nand.num_blocks = flashsize / erasesize;

    flashdata = malloc(flashsize);
    memset(flashdata, 0xff, flashsize);

    uint8_t journal_buf[512];
    struct dhara_map dhara;
    dhara_map_init(&dhara, &nand, journal_buf, gcratio);
    dhara_error_t err = DHARA_E_NONE;
    dhara_map_resume(&dhara, &err);
    printf("Number of physical erase blocks: %d\n", nand.num_blocks);

    uint32_t lba = dhara_map_capacity(&dhara);
    printf("Maximum logical size: %dkB (%d sectors)\n", lba / 2, lba);

    FILE* inf = fopen(inputfilename, "rb");
    if (!inf)
        panic("cannot open input file: %s", strerror(errno));
    fseek(inf, 0, SEEK_END);
    int sectors = ftell(inf) / 512;
    fseek(inf, 0, SEEK_SET);
    if (sectors > lba)
    {
        fprintf(stderr, "Logical image too big (%d > %d)\n",
            sectors, lba);
        exit(1);
    }
    if (sectors == lba)
    {
        fprintf(stderr, "WARNING: logical image completely fills the FTL filesystem; you\n");
        fprintf(stderr, "will have GC thrash if you try to write to it");
    }

    for (int sectorno=0; sectorno<sectors; sectorno++)
    {
        uint8_t buffer[512] = {0};
        fread(buffer, 512, 1, inf);

        err = DHARA_E_NONE;
        dhara_map_write(&dhara, sectorno, buffer, &err); 

        if (err != DHARA_E_NONE)
        {
            fprintf(stderr, "FTL error %d\n", err);
            exit(1);
        }
    }
    fclose(inf);

    err = DHARA_E_NONE;
    dhara_map_gc(&dhara, &err);
    uint32_t used = dhara_map_size(&dhara);
    printf("Used space: %dkB (%d sectors)\n", used / 2, used);

    FILE* outf = fopen(outputfilename, "wb");
    if (!outf)
        panic("cannot open output file: %s", strerror(errno));
    fwrite(flashdata, 1, flashsize, outf);
    fclose(outf);
    return 0;
}

// vim: sw=4 ts=4 et:

